File: VB/CustomLookUpEdit.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Popup
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Controls
Imports System.Reflection
Imports DevExpress.Utils
Imports DevExpress.XtraEditors.ListControls
Imports System.Windows.Forms
Imports System.Collections

Namespace TestLookUpIncrementalSearch
	<UserRepositoryItem("RegisterCustomEdit")>
	Public Class RepositoryItemCustomLookUpEdit
		Inherits RepositoryItemLookUpEdit

		'The static constructor that calls the registration method 
		Shared Sub New()
			RegisterCustomEdit()
		End Sub

		'Initialize new properties 
		Public Sub New()

		End Sub

		'The unique name for the custom editor 
		Public Const CustomEditName As String = "CustomLookUpEdit"

		'Return the unique name 
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return CustomEditName
			End Get
		End Property

		'Register the editor 
        Public Shared Sub RegisterCustomEdit()
            Dim img As Image = Nothing
            EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(CustomEditName, GetType(CustomLookUpEdit), GetType(RepositoryItemCustomLookUpEdit), GetType(LookUpEditViewInfo), New ButtonEditPainter(), True, img))
        End Sub


		'Override the Assign method 
		Public Overrides Sub Assign(ByVal item As RepositoryItem)
			BeginUpdate()
			Try
				MyBase.Assign(item)
				Dim source As RepositoryItemCustomLookUpEdit = TryCast(item, RepositoryItemCustomLookUpEdit)
				If source Is Nothing Then
					Return
				End If
			Finally
				EndUpdate()
			End Try
		End Sub


	End Class

	<ToolboxItem(True)>
	Public Class CustomLookUpEdit
		Inherits LookUpEdit

		'The static constructor that calls the registration method 
		Shared Sub New()
			RepositoryItemCustomLookUpEdit.RegisterCustomEdit()
		End Sub

		'Initialize the new instance 
		Public Sub New()
			'... 
			Me.ViewInfo.UseHighlightSearchAppearance = True
		End Sub

		'Return the unique name 
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemCustomLookUpEdit.CustomEditName
			End Get
		End Property

		Protected Overrides Function CreatePopupForm() As PopupBaseForm
			Return New CustomPopupLookUpEditForm(Me)
		End Function

		'Override the Properties property 
		'Simply type-cast the object to the custom repository item type 
		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)>
		Public Shadows ReadOnly Property Properties() As RepositoryItemCustomLookUpEdit
			Get
				Return TryCast(MyBase.Properties, RepositoryItemCustomLookUpEdit)
			End Get
		End Property

		Private lastChar As String
		Protected Overrides Sub OnPreviewKeyDown(ByVal e As System.Windows.Forms.PreviewKeyDownEventArgs)
			Dim form As CustomPopupLookUpEditForm = TryCast(Me.PopupForm, CustomPopupLookUpEditForm)
			MyBase.OnPreviewKeyDown(e)
            If Char.IsLetterOrDigit(ChrW(e.KeyCode)) Then
                lastChar = ChrW(e.KeyCode).ToString()
            ElseIf e.KeyData = Keys.Back Then
                lastChar = "backspace"
            Else
                lastChar = ""
            End If
		End Sub

		Protected Overrides Sub OnKeyPress(ByVal e As KeyPressEventArgs)
			Dim form As CustomPopupLookUpEditForm = TryCast(Me.PopupForm, CustomPopupLookUpEditForm)
			If lastChar = "backspace" Then
				e.Handled = True
				form.IncrementalSearchText = form.IncrementalSearchText.Remove(form.IncrementalSearchText.Length - 1)
			End If
			MyBase.OnKeyPress(e)
			If lastChar <> "backspace" Then
				form.IncrementalSearchText &= lastChar
			End If
		End Sub

		Public Overrides Sub ClosePopup()
			MyBase.ClosePopup()
		End Sub

	End Class

	Public Class CustomPopupLookUpEditForm
		Inherits PopupLookUpEditForm

		Private _IncrementalSearchText As String = String.Empty
		Public Sub New(ByVal ownerEdit As LookUpEdit)
			MyBase.New(ownerEdit)

		End Sub
		Public Property IncrementalSearchText() As String
			Get
				Return _IncrementalSearchText
			End Get
			Set(ByVal value As String)
				_IncrementalSearchText = value
			End Set
		End Property
		Public Function GetSearchText() As String
			Return Me.SearchText
		End Function

		Protected Overrides Function DoIncrementalSearch(ByVal keyChar As Char, ByVal ctrl As Boolean, ByVal shift As Boolean) As Boolean
			Dim result As Boolean = MyBase.DoIncrementalSearch(keyChar, ctrl, shift)
			selectedRowIndex = Me.SelectedIndex
			Return result
		End Function
		Private selectedRowIndex As Integer = -1

		Protected Overrides Function QueryResultValue() As Object
            Dim _currentValue As String = String.Empty
            Try
                Dim list As IList = TryCast(ListBindingHelper.GetList(Me.OwnerEdit.Properties.DataSource), IList)
                Dim pdc As PropertyDescriptorCollection = ListBindingHelper.GetListItemProperties(Me.OwnerEdit.Properties.DataSource)
                Dim pd As PropertyDescriptor = pdc(Me.OwnerEdit.Properties.Columns(Me.OwnerEdit.Properties.AutoSearchColumnIndex).FieldName)
                _currentValue = pd.GetValue(list(selectedRowIndex)).ToString()
            Catch
            End Try
            If Not Me.OwnerEdit.Properties.CaseSensitiveSearch Then
                _currentValue = _currentValue.ToLower()
                Me.IncrementalSearchText = Me.IncrementalSearchText.ToLower()
            End If
            If Me.IncrementalSearchText = "" OrElse _currentValue.StartsWith(Me.IncrementalSearchText) Then
                Me.IncrementalSearchText = ""
                Return MyBase.QueryResultValue()
            Else
                Me.IncrementalSearchText = ""
                Return MyBase.QueryOldEditValue()
            End If
		End Function


	End Class
End Namespace

File: CS/CustomLookUpEdit.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Popup;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Controls;
using System.Reflection;
using DevExpress.Utils;
using DevExpress.XtraEditors.ListControls;
using System.Windows.Forms;
using System.Collections;

namespace TestLookUpIncrementalSearch {
    [UserRepositoryItem("RegisterCustomEdit")]
    public class RepositoryItemCustomLookUpEdit : RepositoryItemLookUpEdit {

        //The static constructor that calls the registration method 
        static RepositoryItemCustomLookUpEdit() { RegisterCustomEdit(); }

        //Initialize new properties 
        public RepositoryItemCustomLookUpEdit() {

        }

        //The unique name for the custom editor 
        public const string CustomEditName = "CustomLookUpEdit";

        //Return the unique name 
        public override string EditorTypeName { get { return CustomEditName; } }

        //Register the editor 
        public static void RegisterCustomEdit() {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(CustomEditName,
              typeof(CustomLookUpEdit), typeof(RepositoryItemCustomLookUpEdit),
              typeof(LookUpEditViewInfo), new ButtonEditPainter(), true, null));
        }


        //Override the Assign method 
        public override void Assign(RepositoryItem item) {
            BeginUpdate();
            try {
                base.Assign(item);
                RepositoryItemCustomLookUpEdit source = item as RepositoryItemCustomLookUpEdit;
                if(source == null)
                    return;
            } finally {
                EndUpdate();
            }
        }


    }

    [ToolboxItem(true)]
    public class CustomLookUpEdit : LookUpEdit {

        //The static constructor that calls the registration method 
        static CustomLookUpEdit() { RepositoryItemCustomLookUpEdit.RegisterCustomEdit(); }

        //Initialize the new instance 
        public CustomLookUpEdit() {
            //... 
            this.ViewInfo.UseHighlightSearchAppearance = true;
        }

        //Return the unique name 
        public override string EditorTypeName {
            get {
                return
                    RepositoryItemCustomLookUpEdit.CustomEditName;
            }
        }

        protected override PopupBaseForm CreatePopupForm() {
            return new CustomPopupLookUpEditForm(this);
        }

        //Override the Properties property 
        //Simply type-cast the object to the custom repository item type 
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemCustomLookUpEdit Properties {
            get { return base.Properties as RepositoryItemCustomLookUpEdit; }
        }

        string lastChar;
        protected override void OnPreviewKeyDown(System.Windows.Forms.PreviewKeyDownEventArgs e) {
            CustomPopupLookUpEditForm form = this.PopupForm as CustomPopupLookUpEditForm;
            base.OnPreviewKeyDown(e);
            if(char.IsLetterOrDigit((char)e.KeyCode)) {
                lastChar = ((char)e.KeyCode).ToString();
            } else if(e.KeyData == Keys.Back) {
                lastChar = "backspace";
            } else {
                lastChar = "";
            }
        }

        protected override void OnKeyPress(KeyPressEventArgs e) {
            CustomPopupLookUpEditForm form = this.PopupForm as CustomPopupLookUpEditForm;
            if(lastChar == "backspace") {
                e.Handled = true;
                form.IncrementalSearchText = form.IncrementalSearchText.Remove(form.IncrementalSearchText.Length - 1);
            }
            base.OnKeyPress(e);
            if(lastChar != "backspace") {
                form.IncrementalSearchText += lastChar;
            }
        }

        public override void ClosePopup() {
            base.ClosePopup();
        }

    }

    public class CustomPopupLookUpEditForm : PopupLookUpEditForm {
        private string _IncrementalSearchText = String.Empty;
        public CustomPopupLookUpEditForm(LookUpEdit ownerEdit)
            : base(ownerEdit) {

        }
        public string IncrementalSearchText {
            get { return _IncrementalSearchText; }
            set {
                _IncrementalSearchText = value;
            }
        }
        public string GetSearchText() {
            return this.SearchText;
        }

        protected override bool DoIncrementalSearch(char keyChar, bool ctrl, bool shift) {
            bool result = base.DoIncrementalSearch(keyChar, ctrl, shift);
            selectedRowIndex = this.SelectedIndex;
            return result;
        }
        int selectedRowIndex = -1;

        protected override object QueryResultValue() {
            string currentValue = String.Empty;
            try {
                IList list = ListBindingHelper.GetList(this.OwnerEdit.Properties.DataSource) as IList;
                PropertyDescriptorCollection pdc = ListBindingHelper.GetListItemProperties(this.OwnerEdit.Properties.DataSource);
                PropertyDescriptor pd = pdc[this.OwnerEdit.Properties.Columns[this.OwnerEdit.Properties.AutoSearchColumnIndex].FieldName];
                currentValue = pd.GetValue(list[selectedRowIndex]).ToString();
            } catch { }
            if(!this.OwnerEdit.Properties.CaseSensitiveSearch) {
                currentValue = currentValue.ToLower();
                this.IncrementalSearchText = this.IncrementalSearchText.ToLower();
            }
            if(this.IncrementalSearchText == "" || currentValue.StartsWith(this.IncrementalSearchText)) {
                this.IncrementalSearchText = "";
                return base.QueryResultValue();
            } else {
                this.IncrementalSearchText = "";
                return base.QueryOldEditValue();
            }
        }


    }
}
